compiler/rustc_hir/src/definitions.rs

```diff
@@ -337,8 +337,19 @@ impl Definitions {
     }
 
     /// Adds a definition with a parent definition.
-    pub fn create_def(&mut self, parent: LocalDefId, data: DefPathData) -> LocalDefId {
-        debug!("create_def(parent={:?}, data={:?})", parent, data);
+    pub fn create_def(
+        &mut self,
+        crate_name: Symbol,
+        parent: LocalDefId,
+        data: DefPathData,
+    ) -> LocalDefId {
+        // We can't use `Debug` implementation for `LocalDefId` here, since it tries to acquire a
+        // reference to `Definitions` and we're already holding a mutable reference.
+        debug!(
+            "create_def(parent={crate_name}[{short_crate_id:04x}]{local_name}, data={data:?})",
+            short_crate_id = self.stable_crate_id.to_u64() >> 8 * 6,
+            local_name = self.def_path(parent).to_string_no_crate_verbose(),
+        );
```

Comment:
i think we may want to go with just 
```suggestion
        debug!(
            "create_def(parent={}, data={data:?})",
            self.def_path(parent).to_string_no_crate_verbose(),
        );
```
Considering that the crate is always the currently compiled crate it doesn't seem necessary to print that. I don't think manually keeping the printing of the `LocalDefId` in sync with the actual debug impl is feasible long term.

I am not sure what's a great way to warn about future misuses here. Without parralel comp, we get a nice ICE so debugging similar bugs in the future is hopefully not too difficult :thinking:

Reply:
Maybe `Debug()` impl for `LocalDefId` should be removed and replaced by methods like
```
/// Returns a `Debug` representation. If you’re currently holding a mutable reference to `Definitions`, use `.debug_with_definitions()` instead
fn debug(&self) -> impl Debug { ... }
fn debug_with_definitions(&self, definitions: &Definitions) -> impl Debug { ... }
```
?

Reply:
:thinking: i don't think that's worth it. most of the compiler doesn't  use `Definitions` but does want to debug `DefId`s. Removing the `Debug` impl would be very hard, as e.g. `TraitRef` and friends also contain a `DefId`

Reply:
Okay, you’re right.